### PR TITLE
Change parametrization of SmoothlyBrokenPowerLaw to use x_break instead of log_break

### PR DESCRIPTION
**Pull request by astrofrog**
This is to make it consistent with other models and to make it possible to make the model unit-aware. This model is only in dev at the moment, so no deprecation or changelog entry needed, but we should get it in before 2.0

cc @gcalderone (original contributor of this class), @pllim, @nden

### CODE REVIEW

**Comment by gcalderone** [comment]
Moving to x_break may spoil the usefulness of the model, and I bet someone (sooner or later) will re-implement the model with log_break.

The reason is that for a typical usage (fitting of power law spectra) the fit is much more sensitive to small variations in log_break, rather than in x_break.  Moreover, with log_break the parameter uncertainties are usually symmetric, whereas they would be highly asymmetric  with x_break.

This can be appreciated qualitatively by looking at a log-log plot of the model, e.g.:
sbpl

Here log_break = 1.5 ~ 30.  By adding or subtracting 0.5 to log_break we would reach both sides of the plot (10 and 100 respectively). To span the same range with x_break we need to add 70, or subtract 20, hence the asymmetry in the parameter uncertainty.

For further discussion see (e.g.):
http://adsabs.harvard.edu/abs/2015MNRAS.448..403C
http://adsabs.harvard.edu/abs/2007MNRAS.382..342C

The unit-awareness is an issue (currently log_break is supposed to be adimensional), but in my opinion this aspect is less important that the parameter asymmetric behavior.

A viable possibility would be to introduce a layer between the model parametrization presented to the user and the model implementation used by the fitter algorithm.  Such layer may be used to encapsulate both the SmoothlyBrokenPowerLaw (to exhibit the x_break parameter to the user) and the PowerLaw1D model (to re-implement it in terms of log_break avoiding backward incompatibilities).

**Comment by astrofrog** [comment]
@gcalderone - thanks for the feedback. Whatever we do, I think that SmoothlyBrokenPowerLaw should behave the same as BrokenPowerLaw and be parametrized in the same way, which is what motivated the changes here. The issues you describe are not unique to SmoothlyBrokenPowerLaw.

However, I see your point that internally representing the break in log space might make more sense in terms of fitting. I guess one way to do that is to make the official parameter documented in the docstring be x_break and then add a property to the model that automatically converts to/from log space, but I'm not sure if that's the best way to do deal with this.

Just out of curiosity, why is the asymmetry not a problem for the amplitude of the model?

@nden - what are your thoughts on this?

**Comment by astrofrog** [comment]
Just to put the idea out there, is there any way that we could actually generalize the idea of optimizing a parameter in log space, on the fitting side, rather than add special cases like this on a model per model basis? So e.g. saying 'this parameter should be optimized by searching in log space'

**Comment by astrofrog** [comment]
As a side note, I've often had issues fitting power-law models (even plain ones) both in astropy and with scipy directly, and usually always have to revert to doing the whole fitting in log space, so maybe having a way to do that would be useful.

**Comment by gcalderone** [comment]
> Just out of curiosity, why is the asymmetry not a problem for the amplitude of the model?

Actually it is, although to a minor extent.  The reasoning is as follows:

- the normalization of the model is typically constrained by all the available data, while the break energy (or wavelength, or frequency, or whatever...) may be constrained only by a limited portion of the spectrum;

- to provide strong constraints on the break you need reliable data at significantly lower counts (or flux, or luminosity or whatever...) on at least one side of the break.  Here "signficantly" means a factor of 3 at least.

Both this facts imply that the relative uncertainty on the normalization is typically much smaller than 1.  Hence the asymmetry is negligible since d log Norm \propto d Norm.  The same do not applies to x_break since its relative uncertainty is typically >~ 1.

**Comment by gcalderone** [comment]
Setting a flag in the fitter to specify that the quantity to be optimized is the *logarithm* of a parameter (rather than its value) is in my opinion a very good idea.

**Comment by astrofrog** [comment]
@nden - is the idea of optimizing based on the logarithm of a parameter remotely possible given how astropy.modeling works? (I'm not familiar enough with the fitting part to know). If it is, I would suggest merging this as-is and then working on the log fitting for the next release. If however we think it's going to be difficult, then I'd be fine with changing the parameter here back to log but then the BrokenPowerLaw model should be changed to use the log break too (and in both cases I think it should be called ``log_x_break`` rather than ``log_break``. We'll then have to just disallow units on these models.

**Comment by astrofrog** [comment]
Oh wait, for the unit stuff I forgot astropy.units actually supports log units now, so maybe that isn't an issue.

**Comment by nden** [comment]
Implementing optimization of the logarithm of a parameter is possible just not for this release.
We should make an issue for v 3. I hope the log units work.

**Comment by astrofrog** [comment]
@nden - thanks! do you think we should therefore use ``x_break`` or ``log_x_break`` for now for the broken power-law models?

**Comment by nden** [comment]
I am not a user of this model so it's hard to say from that perspective. If you and @gcalderone agree the model is more useful with `log_x_break`, I think a path forward can be to use `log_x_break` in v2 and switch to `x_break` and logarithmic optimization in v3 with the appropriate deprecations. If units don't work with `log_x_break`, skip units implementation in v2 and implement them in v3. How does this sound, @astrofrog and @gcalderone ?

**Comment by gcalderone** [comment]
Switching back and forth from log. parameter definitions across astropy versions is not a good idea (IMHO), so I would suggest to accept this PR and wait for the logarithmic fitter flag to be implemented.

On the other hand, I believe that implementing the possibility to fit a parameter as logarithm may not be straightforward for the following reasons:
- each function/method which handles/calculate uncertainties must be able to cope with asymmetric intervals.  This likely implies doubling inputs/outputs;
- the correlation matrix among parameter is no longer meaningful, at least the one involving the user presented parameter;
- in the future we may need to fit, say, tan(x) but exhibit x to the user.  So we should foresee the possibility to pass the parameter through a "user defined function", rather than force it to be a logarithm;

If all this difficulties prevent us from implementing the logarithmic fitter flag I would definitely choose to maintain the `log_x_break` definition, and possibly re-define the other power law models accordingly.

There is a third solution which is to maintain both the `x_break` and `log_x_break` definitions as separate models.  It is not elegant, but it's very quick and painless.

In summary, my preferences are:
1 - if the "user defined function" feature will be implemented in v2 ==> go to the `x_break` definition;
2 - if the "user defined function" feature will never be implemented ==> maintain the `log_x_break` definition and modify the other powerlaws models;
3 - maintain both the `x_break` and `log_x_break` definitions as separate models.

**Comment by astrofrog** [comment]
My current thinking is that BrokenPowerLaw has been around for a little while, so changing it will require deprecation anyway. But we don't want to do that until we are sure we need to. That is, I don't want to change it and then change it back again in the next version. So the impact would be less on users if we switch SmoothlyBrokenPowerLaw to match (and use ``x_break``). Then if we can figure out a smart way to fit them by v3, we switch to that, and otherwise we change both models to be parametrized by ``log_x_break``. Note that we could still continue to allow initialization via ``x_break`` though, all we would deprecate is accessing ``x_break`` as a parameter potentially, so this would minimize impact on users. There is precedent for 'convenience' initialiation, e.g. in ``Ring2D`` where ``r_out`` can be given but is not a parameter:

```
In [3]: r = Ring2D(x_0=0, y_0=1, r_in=3, r_out=4, width=None)

In [4]: r.r_in
Out[4]: Parameter('r_in', value=3.0)

In [5]: r.r_out
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-ec8eb2ed9625> in <module>()
----> 1 r.r_out

AttributeError: 'Ring2D' object has no attribute 'r_out'
```

In summary, I am suggesting merging this PR, and then before the next version if we can't figure out the log fitting, we change to ``log_x_break`` as the parameter but continue to allow the user to initialize the model with ``x_break``.

**Comment by nden** [comment]
Seems like there's an agreement to keep this as is and implement a functionality allowing parameters to be passed through a function before fitted. @gcalderone's  concerns about the uncertainties and the correlation matrix are valid and difficult to solve.

One last question though. What if the parameter presented to the user is `x_break` while internally it's represented as `log_x_break`? This is currently possible using a setter and getter functions on the Parameter definition (although I'm not sure how the setter and getter will work with units). If we represent it internally as `log_x_break` then that's what's fitted. Would this work?

**Comment by astrofrog** [comment]
@nden - that could work (using the getter and setter) though I haven't experimented enough with that and units to know for sure if it could work.

In any case, I think since we don't want to change BrokenPowerLawModel in a rush now before we really know, the most important is that we merge this PR to make SmoothlyBrokenPowerLawModel at least consistent with BrokenPowerLawModel, and we can experiment with different ways of fixing the fitting in the next version.

**Comment by gcalderone** [comment]
@astrofrog I agree with you that what you propose is the most simple solution.  Still, I believe moving to `x_break` spoils the model.  For my work I will continue to use the logarithmic implementation in any case.

Also the getter/setter approach proposed by @nden may be useful, although we may confuse the user by hiding some unexpected behavior.

In my opinion this discussion clearly points out that there can be several approaches to fit a model, which can not be reduced to a single model implementation without incurring in one or more problems (backward incompatibilities, unexpected behavior, lack of parameter correlation information, fitting optimization, etc...).
Hence this could be the right moment to decide how to cope with different, possibly incompatible, model implementations.

In short, I propose again to implement the `SmoothlyBrokenPowerLaw` in the "standard way" (with `x_break`), and add a new model with the `log_x_break` implementation.  Clearly the former would be a wrapper for the latter, so there would be no code duplication.  I can take care of this implementation.

Doing so the problem would be to find an appropriate name for the new model.  `SmoothlyBrokenPowerLaw_log_break` would be too long?
Any idea?

**Comment by astrofrog** [comment]
If we can avoid too much code duplication, then I agree that the most pragmatic is to have two models for each broken power law model (so just four in total). Maybe LogBrokenPowerLaw and SmoothlyLogBrokenPowerLaw?

Maybe you could open a PR, and we can then make a decision based on how much code this adds? Note that we would need to get this merged by tomorrow evening due to the feature freeze for 2.0. Do you think that's feasible? (this would mean ideally getting the PR done today to leave time for review and testing).

**Comment by gcalderone** [comment]
Ops, I wasn't aware of such time limitation. Frankly, I don't think I could do it within tomorrow.
Given this limitations we are left with two options:
1 - accept your PR and I will implement `SmoothlyBrokenPowerLaw_log_break` for v3;
2 - issue another PR where we simply change the name of the current `SmoothlyBrokenPowerLaw` model to `SmoothlyBrokenPowerLaw_log_break` (or similar), and I will implement the `SmoothlyBrokenPowerLaw` (with `x_break`) within few days.  Doing so there will not be any inconsistency since the `SmoothlyBrokenPowerLaw`  model would no longer be present.

Both options do not introduce backward incompatibilities.  Clearly, I would prefer option 2, but the final decision is up to you.

One final remark: `SmoothlyLogBrokenPowerLaw` seems something different from `SmoothlyBrokenPowerLaw`.  I would prefer adding a suffix...

**Comment by astrofrog** [comment]
@gcalderone - ok, for consistency I will go with your option 1, just so that the two models are consistent in 2.0, and then we can add the two log models for the next version.

**Comment by nden** [comment]
As a side note, the `getter` and `setter` functions will need to be enhanced to work with units, so at the end this was not an option.
